src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/Field.java

```diff
@@ -40,23 +41,79 @@ public class Field {
       fields array (i.e., named, non-VM fields) */
-  Field(InstanceKlass holder, int fieldIndex) {
+  private Field(InstanceKlass holder, int fieldIndex, FieldInfoValues values) {
     this.holder = holder;
     this.fieldIndex = fieldIndex;
+    this.values = values;
+    offset = values.offset;
 
-    offset               = holder.getFieldOffset(fieldIndex);
-    genericSignature     = holder.getFieldGenericSignature(fieldIndex);
-
-    name                 = holder.getFieldName(fieldIndex);
+    name = holder.getSymbolFromIndex(values.nameIndex, isInjected());
+    signature = holder.getSymbolFromIndex(values.signatureIndex, isInjected());
     id          = new NamedFieldIdentifier(name.asString());
-
-    signature            = holder.getFieldSignature(fieldIndex);
     fieldType   = new FieldType(signature);
+    accessFlags = new AccessFlags(values.accessFlags);
+
+    if (isGeneric()) {
+      genericSignature = holder.getSymbolFromIndex(values.genericSignatureIndex, isInjected());
+    }
+  }
 
-    short access         = holder.getFieldAccessFlags(fieldIndex);
-    accessFlags = new AccessFlags(access);
+  /** Constructor for cloning an existing Field object */
+  Field(InstanceKlass holder, int fieldIndex) {
+      this(holder, fieldIndex, holder.getField(fieldIndex).values);
   }
 
+
+  static class FieldInfoValues {
+    int nameIndex;
+    int signatureIndex;
+    int offset;
+    int accessFlags;
+    int fieldFlags;
+    int initialValueIndex;
+    int genericSignatureIndex;
+    int contendedGroup;
+  }
+
+  // The format of the stream, after decompression, is a series of
+  // integers organized like this:
+  //
+  //   FieldInfoStream := j=num_java_fields k=num_injected_fields Field[j+k] End
+  //   Field := name sig offset access flags Optionals(flags)
+  //   Optionals(i) := initval?[i&is_init]     // ConstantValue attr
+  //                   gsig?[i&is_generic]     // signature attr
+  //                   group?[i&is_contended]  // Contended anno (group)
+  //   End = 0
+  //
+
+  static FieldInfoValues readFieldInfoValues(CompressedReadStream crs) {
+    FieldInfoValues fieldInfoValues = new FieldInfoValues();
+    fieldInfoValues.nameIndex = crs.readInt(); // read name_index
+    fieldInfoValues.signatureIndex = crs.readInt(); // read signature index
+    fieldInfoValues.offset = crs.readInt(); // read offset
+    fieldInfoValues.accessFlags = crs.readInt(); // read access flags
+    fieldInfoValues.fieldFlags = crs.readInt();  // read field flags
+    // Optional reads
+    if (fieldIsInitialized(fieldInfoValues.fieldFlags)) fieldInfoValues.initialValueIndex = crs.readInt(); // read initial value index
+    if (fieldIsGeneric(fieldInfoValues.fieldFlags))     fieldInfoValues.genericSignatureIndex = crs.readInt(); // read generic signature index
+    if (fieldIsContended(fieldInfoValues.fieldFlags))   fieldInfoValues.contendedGroup = crs.readInt(); // read contended group
```

Comment:
Column with is too wide. These lines would be easier to read if you made each one multiple lines with curly braces.